src/RecursiveStagesService.php

```diff
@@ -0,0 +1,187 @@
+     *
+     * @param DataObject $object
+     * @param string $stage
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedIdentifiers(DataObject $object, string $stage): array
+    {
+        $identifiers = Versioned::withVersionedMode(function () use ($object, $stage): array {
+            Versioned::set_stage($stage);
+
+            /** @var DataObject $stagedObject */
+            $stagedObject = DataObject::get_by_id($object->ClassName, $object->ID);
+
+            if ($stagedObject === null) {
+                return [];
+            }
+
+            $models = [$stagedObject];
+            $identifiers = [];
+
+            while ($model = array_shift($models)) {
+                // Compose a unique identifier, so we can easily compare models
+                // Note that we intentionally use base class here, so we can cover the situation where model class changes
+                $identifiers[] = implode('_', [$model->baseClass(), $model->ID]);
+                $relatedObjects = $this->getOwnedObjects($model);
+                $models = array_merge($models, $relatedObjects);
+            }
+
+            return $identifiers;
+        });
+
+        sort($identifiers, SORT_STRING);
+
+        return array_values($identifiers);
+    }
+
+    /**
+     * This lookup will attempt to find "owned" objects
+     * This method uses the 'owns' relation, same as @see RecursivePublishable::publishRecursive()
+     *
+     * @param DataObject|RecursivePublishable $object
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedObjects(DataObject $object): array
+    {
+        if (!$object->hasExtension(RecursivePublishable::class)) {
+            return [];
+        }
+
+        // Add versioned stage to cache key to cover the case where non-versioned model owns versioned models
+        // In this situation the versioned models can have different cached state which we need to cover
+        $cacheKey = sprintf('%s-%s', $object->getUniqueKey(), Versioned::get_stage());
+
+        if (!array_key_exists($cacheKey, $this->ownedObjectsCache)) {
+            $this->ownedObjectsCache[$cacheKey] = $object
+                // We intentionally avoid recursive traversal here as it's not memory efficient,
+                // stack based approach is used instead for better performance
+                ->findOwned(false)
+                ->toArray();
+        }
+
+        return $this->ownedObjectsCache[$cacheKey];
+    }
+
+    /**
+     * Determine if model is dirty (has draft changes that need publishing)
+     * Non-versioned models are supported
+     *
+     * @param DataObject $object
+     * @return bool
+     */
+    protected function isModelDirty(DataObject $object): bool
+    {
+        if ($object->hasExtension(Versioned::class)) {
+            /** @var $object Versioned */
+            return !$object->isPublished() || $object->stagesDiffer();
```

Comment:
Why specifically `!isPublished()`? For example take a look at `Versioned::isModifiedOnDraft()`, which according to its PHPDoc "Compares current draft with live version, and returns true if these versions differ, meaning there have been unpublished changes to the draft site."

Seems like we should just be doing this instead?
```suggestion
            return $object->isModifiedOnDraft();
```

Reply:
For that matter, do we need this method at all? The condition for the one place it's called could be `if ($object->hasExtension(Versioned::class) && $object->isModifiedOnDraft())`

Reply:
Good point, will make an update.